sql/core/src/main/scala/org/apache/spark/sql/streaming/TestGroupState.scala

```diff
@@ -0,0 +1,161 @@
+ * // other imports
+ *
+ * // test class setups
+ *
+ * test("Structured Streaming state update function") {
+ *   // Creates the prevState input for the state transition function
+ *   // with desired configs. The create() API would guarantee that
+ *   // the generated instance has the same behavior as the one built by
+ *   // engine with the same configs.
+ *   var prevState = TestGroupState.create[UserStatus](
+ *     optionalState = Optional.empty[UserStatus],
+ *     timeoutConf = EventTimeTimeout,
+ *     batchProcessingTimeMs = 1L,
+ *     eventTimeWatermarkMs = Optional.of(1L),
+ *     hasTimedOut = false)
+ *
+ *   val userId: String = ...
+ *   val actions: Iterator[UserAction] = ...
+ *
+ *   // Asserts the prevState is in init state without updates.
+ *   assert(!prevState.hasUpdated)
+ *
+ *   // Calls the state transition function with the test previous state
+ *   //  with desired configs.
+ *   updateState(userId, actions, prevState)
+ *
+ *   // Asserts the test GroupState object has been updated after calling
+ *   // the state transition function
+ *   assert(prevState.hasUpdated)
+ * }
+ * }}}
+ *
+ * Java example of using `TestGroupSate`:
+ * {{{
+ * import org.apache.spark.sql.streaming.TestGroupState;
+ * // other imports
+ *
+ * // test class setups
+ *
+ * // test `flatMapGroupsWithState` state transition function `updateState()`
+ * public void testUpdateState() {
+ *   // Creates the prevState input for the state transition function
+ *   // with desired configs. The create() API would guarantee that
+ *   // the generated instance has the same behavior as the one built by
+ *   // engine with the same configs.
+ *   TestGroupState prevState = new TestGroupState<UserStatus>().create(
+ *     optionalState = Optional.<UserStatus>empty(),
+ *     timeoutConf = EventTimeTimeout,
+ *     batchProcessingTimeMs = 1L,
+ *     eventTimeWatermarkMs = Optional.of(1L),
+ *     hasTimedOut = false);
+ *
+ *   String userId = ...;
+ *   ArrayList<UserAction> actions = ...;
+ *
+ *   // Asserts the prevState is in init state without updates.
+ *   assertTrue(!prevState.hasUpdated());
+ *
+ *   // Calls the state transition function with the test previous state
+ *   //  with desired configs.
+ *   updateState(userId, actions, prevState);
+ *
+ *   // Asserts the test GroupState object has been updated after calling
+ *   // the state transition function
+ *   assertTrue(prevState.hasUpdated());
+ * }
+ * }}}
+ *
+ * @tparam S User-defined type of the state to be stored for each group. Must be encodable into
+ *           Spark SQL types (see `Encoder` for more details).
+ * @since 3.2.0
+ */
+@Experimental
+@Evolving
+trait TestGroupState[S] extends GroupState[S] {
+  /** Whether the state has been marked for removing */
+  def isRemoved: Boolean
+
+  /** Whether the state has been updated but not removed */
+  def isUpdated: Boolean
+
+  /**
+   * Returns the timestamp if setTimeoutTimestamp is called.
```

Comment:
Link to the corresponding functions in GroupState. If thats complicated, at least put the function names and codee snippets in backticks
```
`setTimeoutTimestamp` 
`Optional.empty`
```